echopop/nwfsc_feat/get_proportions.py

```diff
@@ -1,162 +1,1529 @@
+
+        # If already at target dimensions, return as-is
+        if len(current_dimensions.difference(set(stratify_by))) == 0:
+            return filtered_table
+        else:
+            # Aggregate step by step
+            intermediate_result = filtered_table.unstack("length_bin").sum()
+
+            # Find additional dimensions to unstack
+            remaining_dimensions = list(
+                set(stratify_by).difference(intermediate_result.index.names)
+            )
+
+            # Fill in missing dimensions if needed
+            if len(remaining_dimensions) == 0:
+                remaining_dimensions = list(set(stratify_by).difference(["length_bin"]))
+
+            # Final unstack
+            return intermediate_result.unstack(remaining_dimensions)
+
+
+def get_weight_proportions_slice(
+    weight_proportions: pd.DataFrame,
+    stratify_by: List[str],
+    include_filter: Dict[str, Any] = {},
+    number_proportions: Optional[Union[Dict[str, pd.DataFrame], pd.DataFrame]] = None,
+    length_threshold_min: float = 0.0,
+    weight_proportion_threshold: float = 1e-10,
+) -> pd.Series:
+    """
+    Calculate weight proportions for a population slice with optional thresholding.
+
+    This function computes weight proportions for a target population group, with
+    optional thresholding based on number proportions. The thresholding helps
+    identify cases where estimates may be unreliable due to low sample sizes.
+
+    Parameters
+    ----------
+    weight_proportions : pd.DataFrame
+        DataFrame with weight proportions and hierarchical index structure
+    stratify_by : List[str]
+        Stratification columns (e.g., ["stratum_ks"])
+    include_filter : Dict[str, Any], default {}
+        Filter criteria for target group, e.g., {"age_bin": [1]}
+    number_proportions : Union[Dict[str, pd.DataFrame], pd.DataFrame], optional
+        Number proportions for thresholding calculation:
+        - Dict: multiple datasets (e.g., {"aged": df1, "unaged": df2})
+        - DataFrame: single dataset
+        - None: no thresholding applied
+    length_threshold_min : float, default 0.0
+        Minimum length for threshold calculations (e.g., 10.0 cm)
+        Only used when number_proportions is provided
+    weight_proportion_threshold : float, default 1e-10
+        Threshold value for proportion comparisons
+
+    Returns
+    -------
+    pd.Series
+        Weight proportions by strata for the target group
+
+    Notes
+    -----
+    Thresholding Logic:
+    1. Calculate target group weight proportions normally
+    2. If number_proportions provided, calculate corresponding number thresholds
+    3. Apply threshold mask where both weight and number proportions are very small
+    4. Set masked values to 0.0 to indicate unreliable estimates
+
+    The function handles dictionary inputs by computing element-wise products
+    across all datasets to identify cases where multiple sources agree on
+    very low proportions.
```

Comment:
Not sure which are the "dictionary inputs" you refer to here. What are the multiple sources? and which proportion are you talking about here when you say "low proportions" ?

I think try to write in a more concrete way, like just write out what will the function do when X source is available, when Y source is available, and when X and Y sources are both available, will be helpful to users.

Since this is very specific to FEATs hake workflow, I wouldn't worry about writing something generic. The operations and inputs/outputs are already very locked in. This is a very different scenario compared to the inversion classes.